helpers/pedigree_from_sample_metadata.py

```diff
@@ -0,0 +1,227 @@
+    """
+    fetches the project pedigree from sample-metadata
+    list, one dict per participant
+    :param project:
+    :return: all content retrieved from API
+    """
+
+    return FamilyApi().get_pedigree(project=project)
+
+
+def ext_to_int_sample_map(project: str) -> Dict[str, List[str]]:
+    """
+    fetches the participant-sample mapping, so external IDs can be translated
+    to the corresponding CPG ID
+
+    This endpoint returns a list of tuples, each containing a participant ID
+    and corresponding sample ID
+
+    This originally had an expectation of a 1:1 participant:sample relationship
+    that will not hold in general, as numerous samples have multiple samples
+
+    for each participant, create a list of all possible samples
+
+    :param project:
+    :return: the mapping dictionary
+    """
+
+    sample_map = defaultdict(list)
+    for (
+        participant,
+        sample,
+    ) in ParticipantApi().get_external_participant_id_to_internal_sample_id(
+        project=project
+    ):
+        sample_map[participant].append(sample)
+    return sample_map
+
+
+def generate_reverse_lookup(mapping_digest: Dict[str, List[str]]) -> Dict[str, str]:
+    """
+    :param mapping_digest: created by ext_to_int_sample_map
+    :return:
+    """
+
+    return {
+        sample: participant
+        for participant, samples in mapping_digest.items()
+        for sample in samples
+    }
+
+
+@click.command()
+@click.option(
+    '--project',
+    help='the name of the project to use in API queries',
+)
+@click.option(
+    '--singletons',
+    default=False,
+    is_flag=True,
+    help='remake the pedigree as singletons',
+)
+@click.option(
+    '--plink',
+    default=False,
+    is_flag=True,
+    help='make a plink format file (.fam, .ped is the default)',
+)
+@click.option(
+    '--output',
+    help='prefix for writing all outputs to',
+)
+def main(project: str, singletons: bool, plink: bool, output: str):
+    """
+
+    :param project: may be able to retrieve this from the environment
+    :param singletons: whether to split the pedigree(s) into singletons
+    :param plink: whether to write the file as PLINK.fam format
+    :param output: path to write new PED file
+    """
+
+    # get the list of all pedigree members as list of dictionaries
+    logging.info('Pulling all pedigree members')
+    pedigree_dicts = get_pedigree_for_project(project=project)
+
+    # endpoint gives list of tuples e.g. [['A1234567_proband', 'CPG12341']]
+    # parser returns a dictionary, arbitrary # sample IDs per participant
+    logging.info('pulling internal-external sample mapping')
+    sample_to_cpg_dict = ext_to_int_sample_map(project=project)
+
+    # store a way of reversing this lookup in future
+    reverse_lookup = generate_reverse_lookup(sample_to_cpg_dict)
+    with open(f'{output}_reversed.json', 'w', encoding='utf-8') as handle:
+        json.dump(reverse_lookup, handle, indent=4)
```

Comment:
The main reason is to separate the generation of the PED/lookup file from the analysis process. In another setting the Pedigree/lookup could come from a source other than the SampleMetadata API, so that shouldn't be a limiting factor in running AIP. In those other settings it's also possible that the Sample ID is consistent, so there's no need to do this looking up between internal and external, so this is only an optional argument later on